### PR TITLE
Use kafka-streams-framework

### PR DESCRIPTION
**Pull request by surajpuvvada**
also check if KStream for an input topic already exists and re-use it.

### CODE REVIEW

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java [hunk 1]
```diff
@@ -1,58 +1,121 @@
...
+import org.hypertrace.core.serviceframework.config.ConfigUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-/**
- * ViewGeneratorLauncher reads enriched structured traces and converts them to Pinot records using specified Avro
- * schema specified in the target view generator class, which is specified in the config.
- */
-public class ViewGeneratorLauncher extends PlatformService {
+public class ViewGeneratorLauncher extends KafkaStreamsApp {
```
Comment (tim-mwangi) [inline view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java:28]: Can kafka streams framework base class extend PlatformService or is it already doing that?

**Comment by kotharironak** [comment]
@surajpuvvada @laxmanchekka Last commits update the PR with latest framework changes and configure configs accordingly.

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java [hunk 2]
```diff
@@ -1,58 +1,108 @@
...
-  private StreamExecutionEnvironment environment;
+  private static final Logger logger = LoggerFactory.getLogger(ViewGeneratorLauncher.class);
+  private static final String DEFAULT_VIEW_GEN_JOB_CONFIG_KEY = "view-gen";
+  private String viewGenName;
 
   public ViewGeneratorLauncher(ConfigClient configClient) {
     super(configClient);
   }
 
+  public String getViewGenName() {
```
Comment (tim-mwangi) [inline view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java:34]: What are these get and set methods used for?

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java [hunk 3]
```diff
@@ -1,58 +1,108 @@
...
+    return viewGenName;
+  }
+
+  public void setViewGenName(String viewGenName) {
+    this.viewGenName = viewGenName;
+  }
+
   @Override
-  protected void doInit() {
+  public StreamsBuilder buildTopology(Map<String, Object> properties, StreamsBuilder streamsBuilder,
```
Comment (tim-mwangi) [inline view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java:43]: So I've been seeing this method and others overridden methods in other extensions of KafkaStreamsApp. To me it seems they can be pushed into the superclass and anything specific to the extension can be passed in as a parameter.

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java [hunk 4]
```diff
@@ -1,58 +1,108 @@
...
+      viewMapper = new InputToViewMapper(viewGeneratorClassName);
     } catch (Exception e) {
-      LOGGER.error("Failed to initialize ViewGenerationJob.", e);
       throw new RuntimeException(e);
     }
+
+    String inputTopic = jobConfig.getString(INPUT_TOPIC_CONFIG_KEY);
+    String outputTopic = jobConfig.getString(OUTPUT_TOPIC_CONFIG_KEY);
+
+    KStream<?, ?> inputStream = inputStreams.get(inputTopic);
```
Comment (tim-mwangi) [inline view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java:59]: Can we avoid wildcard generics if we can?

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java (hunk 1 above)
Comment (surajpuvvada) [reply view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java:24]: yes the KafkaStreamsApp extends PlatformService

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java (hunk 2 above)
Comment (kotharironak) [reply view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java:34]: It is used in `MultiViewGeneratorLauncher` where all view-generator are merged as a single KStream job. In that case,  getViewGenName is used from getJobConfigKey which help to fetch its specific config from properties.

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorConstants.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package org.hypertrace.core.viewgenerator.service;
+
+public class ViewGeneratorConstants {
+  public static final String SCHEMA_REGISTRY_CONFIG_KEY = "schema.registry.config";
```
Comment (laxmanchekka) [inline view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorConstants.java:4]: Unused?

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java [hunk 6]
```diff
@@ -1,58 +1,108 @@
...
-    try {
-      environment.execute(getServiceName());
-    } catch (Exception e) {
-      LOGGER.error("Got exception during ViewGenerationJob.", e);
-      // Since the job couldn't recover from the error state like this.
-      // It's better to kill the entire process and let the guardian process to restart the job,
-      // e.g. systemd or kubernetes.
-      System.exit(1);
-    }
+  public Map<String, Object> getStreamsConfig(Config config) {
```
Comment (laxmanchekka) [inline view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java:76]: Duplicate code as in super class?

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java (hunk 3 above)
Comment (kotharironak) [reply view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java:43]: we did this by providing a default implementation of most of the interface methods except buildTopology. Get[Input/Output]Topics impl will be needed in MultiViewGen and Hypertrace-ingester.

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java (hunk 6 above)
Comment (kotharironak) [reply view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java]: Done

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java (hunk 4 above)
Comment (kotharironak) [reply view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java:59]: we would like to merge our ingester pipeline as one stream in hypertrace-ingester. Here, we will have different input streams types with RawSpan, StructuredTrace, so I am not very sure that we can base templatize or have the interface with Object.  @surajpuvvada  what do you think? The change will need an interface change of buildTopology, so, I would suggest doing as follow up to finish with the first e2e pipeline.

File: view-generator-framework/build.gradle.kts [hunk 7]
```diff
@@ -35,14 +35,7 @@ dependencies {
...
 
-  // Flink
-  implementation("org.apache.flink:flink-streaming-java_2.11:1.7.0")
-  implementation("org.apache.flink:flink-connector-kafka_2.11:1.7.0")
-  implementation("de.javakaffee:kryo-serializers:0.45")
-  runtimeOnly("org.apache.flink:flink-avro:1.9.2")
-
-  // Needed for flink metric exporter. Used for Hypertrace and debugging.
-  runtimeOnly("org.apache.flink:flink-metrics-slf4j:1.10.1")
+  implementation("org.hypertrace.core.kafkastreams.framework:kafka-streams-framework:0.1.7")
```
Comment (surajpuvvada) [inline view-generator-framework/build.gradle.kts:38]:
```suggestion
  implementation("org.hypertrace.core.kafkastreams.framework:kafka-streams-framework:0.1.8")
```

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java (hunk 3 above)
Comment (surajpuvvada) [reply view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java:43]: @tim-mwangi most of the jobs have a different impl for this method - what do you feel can be moved to base class ?

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java (hunk 4 above)
Comment (surajpuvvada) [reply view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java:59]: yes the variables could be of multiple types @tim-mwangi do you have any suggestions ?

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java (hunk 3 above)
Comment (tim-mwangi) [reply view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/ViewGeneratorLauncher.java:43]: ok. I was just curious if there's code duplication.

**Review by surajpuvvada — COMMENTED** [review]
@kotharironak lgtm. 
@laxmanchekka do u have any pending changes to push ?

File: view-generator-framework/build.gradle.kts [hunk 8]
```diff
@@ -35,14 +35,7 @@ dependencies {
...
 
-  // Flink
-  implementation("org.apache.flink:flink-streaming-java_2.11:1.7.0")
-  implementation("org.apache.flink:flink-connector-kafka_2.11:1.7.0")
-  implementation("de.javakaffee:kryo-serializers:0.45")
-  runtimeOnly("org.apache.flink:flink-avro:1.9.2")
-
-  // Needed for flink metric exporter. Used for Hypertrace and debugging.
-  runtimeOnly("org.apache.flink:flink-metrics-slf4j:1.10.1")
+  implementation("org.hypertrace.core.kafkastreams.framework:kafka-streams-framework:0.1.8")
```
Comment (kotharironak) [inline view-generator-framework/build.gradle.kts:38]:
```suggestion
  implementation("org.hypertrace.core.kafkastreams.framework:kafka-streams-framework:0.1.9")
```

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/MultiViewGeneratorLauncher.java [hunk 9]
```diff
@@ -1,87 +1,113 @@
...
-    try {
-      environment.execute(getServiceName());
-    } catch (Exception e) {
-      LOGGER.error("Error occurred in view generation", e);
-      // Since the job couldn't recover from the error state like this.
-      // It's better to kill the entire process and let the guardian process to restart the job,
-      // e.g. systemd or kubernetes.
-      System.exit(1);
-    }
+  public Logger getLogger() {
```
Comment (laxmanchekka) [inline view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/MultiViewGeneratorLauncher.java:61]: Pls remove this

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/MultiViewGeneratorLauncher.java [hunk 10]
```diff
@@ -1,87 +1,113 @@
...
+      Config viewGenConfig = viewGenConfigs.get(viewGen);
+      inputTopics.add(viewGenConfig.getString(INPUT_TOPIC_CONFIG_KEY));
+    }
+    return inputTopics.stream().collect(Collectors.toList());
   }
 
+  @Override
+  public List<String> getOutputTopics(Map<String, Object> properties) {
+    List<String> viewGenNames = getViewGenName(properties);
+    Set<String> inputTopics = new HashSet<>();
```
Comment (laxmanchekka) [inline view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/MultiViewGeneratorLauncher.java:84]: Variable name - outputtopics?

**Review by laxmanchekka — APPROVED** [review]

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/MultiViewGeneratorLauncher.java (hunk 9 above)
Comment (kotharironak) [reply view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/MultiViewGeneratorLauncher.java]: done

File: view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/MultiViewGeneratorLauncher.java (hunk 10 above)
Comment (kotharironak) [reply view-generator-framework/src/main/java/org/hypertrace/core/viewgenerator/service/MultiViewGeneratorLauncher.java]: done

**Review by laxmanchekka — APPROVED** [review]

**Pull request merged**